File: Skill.cpp
#include "Skill.hpp"

using namespace std;

Skill :: Skill(string name, int basePower)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 0;
	listOfSkill.push_back(*this);
}

Skill :: Skill(string name, int basePower, Element e)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 1;
	this->elements.push_back(e);
	listOfSkill.push_back(*this);
}

Skill :: Skill(string name, int basePower, Element e1, Element e2)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 2;
	this->elements.push_back(e1);
	this->elements.push_back(e2);
	listOfSkill.push_back(*this);
}

Skill :: Skill(string name, int basePower, Element e1, Element e2, Element e3)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 3;
	this->elements.push_back(e1);
	this->elements.push_back(e2);
	this->elements.push_back(e3);
	listOfSkill.push_back(*this);
}

Skill :: Skill(string name, int basePower, Element e1, Element e2, Element e3, Element e4)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 4;
	this->elements.push_back(e1);
	this->elements.push_back(e2);
	this->elements.push_back(e3);
	this->elements.push_back(e4);
	listOfSkill.push_back(*this);
}

Skill :: Skill(string name, int basePower, Element e1, Element e2, Element e3, Element e4, Element e5)
{
	this->skillName = name;
	this->basePower = basePower;
	this->masteryLevel = 1;
	this->numOfElements = 5;
	this->elements.push_back(e1);
	this->elements.push_back(e2);
	this->elements.push_back(e3);
	this->elements.push_back(e4);
	this->elements.push_back(e5);
	listOfSkill.push_back(*this);
}

Skill :: Skill(const Skill& s) : skillName(s.getName())
{
	this->basePower = s.getBasePower();
	this->masteryLevel = s.getMasteryLevel();
	this->numOfElements = s.getNumOfElements();
	for(int i = 0; i < this->numOfElements; i++)
	{
		this->elements.push_back(s.getElement(i));
	}
}

void Skill :: operator=(const Skill& s)
{
	this->skillName = s.getName();
	this->basePower = s.getBasePower();
	this->masteryLevel = s.getMasteryLevel();
	this->numOfElements = s.getNumOfElements();
	for(int i = 0; i < this->numOfElements; i++)
	{
		this->elements.push_back(s.getElement(i));
	}
}

bool Skill :: operator==(const Skill& s)
{
	return this->skillName == s.getName();
}

Skill :: ~Skill()
{
}

ostream& operator<<(ostream& out, const Skill& s)
{
	out << s.getName() << endl;
	out << "Base Power   : " << s.getBasePower() << endl;
	out << "Mastery Level: " << s.getMasteryLevel() << endl;
	/*
	out << "Can be learnt by " << s.getNumOfElements() << " element(s):" << endl;
	
	for(int i = 0; i < s.getNumOfElements(); i++)
	{
		out << i + 1 << ". " << s.getElement(i) << endl;
	}
	*/
	return out;
}

string Skill :: getName() const
{
	return this->skillName;
}
		
int Skill :: getBasePower() const
{
	return this->basePower;
}

int Skill :: getMasteryLevel() const
{
	return this->masteryLevel;
}

void Skill :: setMasteryLevel(int newMasteryLevel)
{
	this->masteryLevel = newMasteryLevel;
}

int Skill :: getNumOfElements() const
{
	return this->numOfElements;
}


vector<Element> Skill :: getElements() const {
	return this->elements;
}
		
bool Skill :: isLearntBy(Element e) const
{
	for(int i = 0; i < this->numOfElements; i++)
	{
		if(this->elements[i] == e) return true;
	}
	return false;
}

Element Skill :: getElement(int index) const
{
	if(index >= this->numOfElements || index < 0)
	{
		throw IndexInvalidException();
	}
	
	return this->elements[index];
}

Skill& getSkillByName(string name)
{
	for(int i = 0; i < listOfSkill.size(); i++)
	{
		if(listOfSkill[i].getName() == name) return listOfSkill[i];
	}
	throw ItemNotFoundException();
}

vector<Skill> listOfSkill;

Skill Tackle("Tackle", 50, Fire, Ice, Ground, Electric, Water);
Skill BodySlam("Body Slam", 90, Fire, Ice, Ground, Electric, Water);
Skill DoubleEdge("Double Edge", 120, Fire, Ice, Ground, Electric, Water);
Skill GigaImpact("Giga Impact", 150, Fire, Ice, Ground, Electric, Water);
Skill FirePledge("Fire Pledge", 50, Fire);
Skill FlameThrower("Flamethrower", 200, Fire);
Skill FireBlast("Fire Blast", 120, Fire);
Skill BlastBurn("Blast Burn", 150, Fire);
Skill Avalanche("Avalanche", 80, Ice);
Skill IceBeam("Ice Beam", 50, Ice);
Skill Blizzard("Blizzard", 120, Ice);
Skill SubzeroSlammer("Subzero Slammer", 140, Ice);
Skill BullDoze("BullDoze", 60, Ground);
Skill ThousandArrows("Thousand Arrows", 90, Ground);
Skill Earthquake("Earthquake", 100, Ground);
Skill TectonicRage("Tectonic Rage", 160, Ground);
Skill ShockWave("Shock Wave", 60, Electric);
Skill Thunderbolt("Thunderbolt", 90, Electric);
Skill VoltTackle("Volt Tackle", 120, Electric);
Skill GigavoltHavoc("Gigavolt Havoc", 160, Electric);
Skill WaterPledge("Water Pledge", 50, Water);
Skill Surf("Surf", 90, Water);
Skill HydroPump("Hydro Pump", 120, Water);
Skill HydroCannon("Hydro Cannon", 150, Water);

Skill Liquidation("Liquidation", 85, Water, Ice);
Skill IceHammer("Ice Hammer", 100, Ice, Water);
Skill MaxHailstorm("Max Hailstorm", 110, Ground, Ice);
Skill IcicleSpear("Icicle Spear", 75, Ice, Ground);
Skill FreezeJolt("Freeze Jolt", 105, Ice, Electric);
Skill QuadrupleAxel("Quadruple Axel", 80, Electric, Ice);
Skill FreezingFlame("Freezing Flame", 110, Ice, Fire);
Skill BurningChill("Burning Chill", 75, Fire, Ice);
Skill Thunder("Thunder", 110, Electric, Water);
Skill Storm("Storm", 75, Water, Electric);
Skill Mudflood("Mudflood", 95, Ground, Water);
Skill MuddyWater("Muddy Water", 90, Water, Ground);
Skill SteamBlast("Steam Blast", 105, Water, Fire);
Skill SteamBomb("Steam Bomb", 80, Fire, Water);
Skill ThunderBlast("Thunder Blast", 110, Electric, Ground);
Skill EarthyShock("Earthy Shock", 75, Ground, Electric);
Skill ShockAndBurn("Shock and Burn", 100, Electric, Fire);
Skill BurnAndShock("Burn and Shock", 85, Fire, Electric);
Skill Eruption("Eruption", 100, Fire, Ground);
Skill LavaPlume("Lava Plume", 85, Fire, Ground);

Skill WeatherBall("Weather Ball", 90, Fire, Water, Ice);
Skill TriAttack("Tri Attack", 90, Fire, Ice, Electric);
Skill EarthPower("Earth Power", 90, Fire, Ground, Water);
Skill Korslet("Korslet", 90, Fire, Electric, Water);
Skill MeltTheGround("Melt the Ground", 90, Fire, Ground, Ice);
Skill ShockTheFlame("Shock the Flame", 90, Fire, Ground, Electric);
Skill FreezeDry("Freeze Dry", 90, Ice, Water, Ground);
Skill Electrolysis("Electrolysis", 90, Ice, Water, Electric);
Skill ColdRefrigerator("Cold Refrigerator", 90, Ice, Electric, Ground);
Skill ContradictingShock("Contradicting Shock", 90, Ground, Water, Electric);
Skill GodlyFart("GodlyFart", 90, Fire, Ice, Ground, Electric, Water);

File: Skill.java
import java.util.ArrayList;

public class Skill implements Comparable<Skill> {
    // atribut Skill
    public final String skillName;
    public final int basePower;
    private byte masteryLevel;
    public final ArrayList<Element> learnableBy;

    // menampung semua Skill yang dienumerasi di bawah
    public static ArrayList<Skill> listOfSkill = new ArrayList<>();

    // constructor di bawah ini diprivate agar tidak ada Skill baru oleh pengguna
    private Skill(String skillName, int basePower, ArrayList<Element> e) {
        this.skillName = skillName;
        this.basePower = basePower;
        this.masteryLevel = 1;
        this.learnableBy = e;
        Skill.listOfSkill.add(this);
    }

    // hanya copy constructor yang boleh dipakai
    public Skill(Skill s) {
        this.skillName = s.skillName;
        this.basePower = s.basePower;
        this.masteryLevel = s.getMasteryLevel();
        this.learnableBy = new ArrayList<>(s.learnableBy);
        this.learnableBy.sort(Element.comparator);
    }

    public void increaseMasteryLevel() {
        if (this.masteryLevel == 3)
            return;
        this.masteryLevel++;
    }

    public void setMasteryLevel(byte masteryLevel) {
        this.masteryLevel = masteryLevel;

        if (this.masteryLevel <= 1)
            this.masteryLevel = 1;
        if (this.masteryLevel >= 3)
            this.masteryLevel = 3;
    }

    public byte getMasteryLevel() {
        return this.masteryLevel;
    }

    public String toString() {
        String s = "";
        s = s + this.skillName + "\n";
        s = s + "Power : " + this.basePower + "\n";
        s = s + "Mastery : " + this.masteryLevel;
        return s;
    }

    public boolean equals(Skill s) {
        return this.skillName.equals(s.skillName);
    }

    public static Skill getBySkillName(String skillName) {
        for (Skill s : Skill.listOfSkill) {
            if (s.skillName.equals(skillName))
                return s;
        }
        return null;
    }

    public int compareTo(Skill s) {
        return this.basePower - s.basePower;
    }

    // daftar Skill (dienumerasi)
    public static final Skill TACKLE = new Skill("Tackle", 50,
            Element.constructElements("FIRE, ICE, GROUND, ELECTRIC, WATER"));
    public static final Skill BODYSLAM = new Skill("Body Slam", 90,
            Element.constructElements("FIRE, ICE, GROUND, ELECTRIC, WATER"));
    public static final Skill DOUBLEEDGE = new Skill("Double Edge", 120,
            Element.constructElements("FIRE, ICE, GROUND, ELECTRIC, WATER"));
    public static final Skill GIGAIMPACT = new Skill("Giga Impact", 150,
            Element.constructElements("FIRE, ICE, GROUND, ELECTRIC, WATER"));
    public static final Skill FIREPLEDGE = new Skill("Fire Pledge", 50, Element.constructElements("FIRE"));
    public static final Skill FLAMETHROWER = new Skill("Flamethrower", 90, Element.constructElements("FIRE"));
    public static final Skill FIREBLAST = new Skill("Fire Blast", 120, Element.constructElements("FIRE"));
    public static final Skill BLASTBURN = new Skill("Blast Burn", 150, Element.constructElements("FIRE"));
    public static final Skill AVALANCHE = new Skill("Avalanche", 60, Element.constructElements("ICE"));
    public static final Skill ICEBEAM = new Skill("Ice Beam", 90, Element.constructElements("ICE"));
    public static final Skill BLIZZARD = new Skill("Blizzard", 120, Element.constructElements("ICE"));
    public static final Skill SUBZEROSLAMMER = new Skill("Subzero Slammer", 140, Element.constructElements("ICE"));
    public static final Skill BULLDOZE = new Skill("BullDoze", 60, Element.constructElements("GROUND"));
    public static final Skill THOUSANDARROWS = new Skill("Thousand Arrows", 90, Element.constructElements("GROUND"));
    public static final Skill EARTHQUAKE = new Skill("Earthquake", 100, Element.constructElements("GROUND"));
    public static final Skill TECTONICRAGE = new Skill("Tectonic Rage", 160, Element.constructElements("GROUND"));
    public static final Skill SHOCKWAVE = new Skill("Shock Wave", 60, Element.constructElements("ELECTRIC"));
    public static final Skill THUNDERBOLT = new Skill("Thunderbolt", 90, Element.constructElements("ELECTRIC"));
    public static final Skill VOLTTACKLE = new Skill("Volt Tackle", 120, Element.constructElements("ELECTRIC"));
    public static final Skill GIGAVOLTHAVOC = new Skill("Gigavolt Havoc", 140, Element.constructElements("ELECTRIC"));
    public static final Skill WATERPLEDGE = new Skill("Water Pledge", 50, Element.constructElements("WATER"));
    public static final Skill SURF = new Skill("Surf", 90, Element.constructElements("WATER"));
    public static final Skill HYDROPUMP = new Skill("Hydro Pump", 120, Element.constructElements("WATER"));
    public static final Skill HYDROCANNON = new Skill("Hydro Cannon", 150, Element.constructElements("WATER"));
    public static final Skill LIQUIDATION = new Skill("Liquidation", 85, Element.constructElements("WATER, ICE"));
    public static final Skill ICEHAMMER = new Skill("Ice Hammer", 100, Element.constructElements("ICE, WATER"));
    public static final Skill MAXHAILSTORM = new Skill("Max Hailstorm", 110, Element.constructElements("GROUND, ICE"));
    public static final Skill ICICLESPEAR = new Skill("Icicle Spear", 75, Element.constructElements("ICE, GROUND"));
    public static final Skill FREEZEJOLT = new Skill("Freeze Jolt", 105, Element.constructElements("ICE, ELECTRIC"));
    public static final Skill QUADRUPLEAXEL = new Skill("Quadruple Axel", 80,
            Element.constructElements("ELECTRIC, ICE"));
    public static final Skill FREEZINGFLAME = new Skill("Freezing Flame", 110, Element.constructElements("ICE, FIRE"));
    public static final Skill BURNINGCHILL = new Skill("Burning Chill", 75, Element.constructElements("FIRE, ICE"));
    public static final Skill THUNDER = new Skill("Thunder", 110, Element.constructElements("ELECTRIC, WATER"));
    public static final Skill STORM = new Skill("Storm", 75, Element.constructElements("WATER, ELECTRIC"));
    public static final Skill MUDFLOOD = new Skill("Mudflood", 95, Element.constructElements("GROUND, WATER"));
    public static final Skill MUDDYWATER = new Skill("Muddy Water", 90, Element.constructElements("WATER, GROUND"));
    public static final Skill STEAMBLAST = new Skill("Steam Blast", 105, Element.constructElements("WATER, WATER"));
    public static final Skill STEAMBOMB = new Skill("Steam Bomb", 80, Element.constructElements("FIRE, WATER"));
    public static final Skill THUNDERBLAST = new Skill("Thunder Blast", 110,
            Element.constructElements("ELECTRIC, GROUND"));
    public static final Skill EARTHYSHOCK = new Skill("Earthy Shock", 75,
            Element.constructElements("GROUND, ELECTRIC"));
    public static final Skill SHOCKANDBURN = new Skill("Shock and Burn", 100,
            Element.constructElements("ELECTRIC, FIRE"));
    public static final Skill BURNANDSHOCK = new Skill("Burn and Shock", 85,
            Element.constructElements("FIRE, ELECTRIC"));
    public static final Skill ERUPTION = new Skill("Eruption", 100, Element.constructElements("FIRE, GROUND"));
    public static final Skill LAVAPLUME = new Skill("Lava Plume", 85, Element.constructElements("FIRE, GROUND"));
    public static final Skill WEATHERBALL = new Skill("Weather Ball", 90,
            Element.constructElements("FIRE, WATER, ICE"));
    public static final Skill TRIATTACK = new Skill("Tri Attack", 90, Element.constructElements("FIRE, ICE, ELECTRIC"));
    public static final Skill EARTHPOWER = new Skill("Earth Power", 90,
            Element.constructElements("FIRE, GROUND, WATER"));
    public static final Skill KORSLET = new Skill("Korslet", 90, Element.constructElements("FIRE, ELECTRIC, WATER"));
    public static final Skill MELTTHEGROUND = new Skill("Melt the Ground", 90,
            Element.constructElements("FIRE, GROUND, ICE"));
    public static final Skill SHOCKTHEFLAME = new Skill("Shock the Flame", 90,
            Element.constructElements("FIRE, GROUND, ELECTRIC"));
    public static final Skill FREEZEDRY = new Skill("Freeze Dry", 90, Element.constructElements("ICE, WATER, GROUND"));
    public static final Skill ELECTROLYSIS = new Skill("Electrolysis", 90,
            Element.constructElements("ICE, WATER, ELECTRIC"));
    public static final Skill COLDREFRIGERATOR = new Skill("Cold Refrigerator", 90,
            Element.constructElements("ICE, ELECTRIC, GROUND"));
    public static final Skill CONTRADICTINGSHOCK = new Skill("Contradicting Shock", 90,
            Element.constructElements("GROUND, WATER, ELECTRIC"));
    public static final Skill GODLYFART = new Skill("Godly Fart", 90,
            Element.constructElements("FIRE, ICE, GROUND, ELECTRIC, WATER"));

}
